// @softwareventures/ordered/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.reverse = exports.compare = exports.Comparison = void 0;
/**
 * The result of a comparison between two ordered values.
 */
var Comparison;
(function (Comparison) {
    Comparison[Comparison["before"] = -1] = "before";
    Comparison[Comparison["equal"] = 0] = "equal";
    Comparison[Comparison["after"] = 1] = "after";
    Comparison[Comparison["undefined"] = NaN] = "undefined";
})(Comparison = exports.Comparison || (exports.Comparison = {}));
/**
 * A general purpose comparator for comparing primitive values.
 */
const compare = (a, b) => {
    if (a < b) {
        return Comparison.before;
    }
    else if (a === b) {
        return Comparison.equal;
    }
    else if (a > b) {
        return Comparison.after;
    }
    else {
        return Comparison.undefined;
    }
};
exports.compare = compare;
/**
 * Returns a new comparator that gives opposite results to the provided
 * comparator.
 */
function reverse(comparator) {
    return (a, b) => -comparator(a, b);
}
exports.reverse = reverse;
